### PR TITLE
bpo-31940: Reject faulty lchmod implementations

### PR DESCRIPTION
**Pull request by asottile**
This is a runtime attempt at implementing https://github.com/python/cpython/pull/4267

A few notes:
- During review it was suggested to _cache_ the failed `lchmod` attempts, however I don't think this is doable:
    - `lchmod` is used for both symlinks and not symlinks if `follow_symlinks=False` -- it works fine for normal files
    - detecting whether something is a symlink up front throws away all benefits of caching whether it fails as you still make a syscall
    - the code for `fchmodat` adjacent to the `lchmod` code serves as precedent for both of these decisions

https://bugs.python.org/issue31940

### CODE REVIEW

**Comment by fruch** [comment]
@asottile nice, can't understand how there's a coredump in the test_shutil.py (under clang). 
you tested this one only under alpine ?

**Comment by asottile** [comment]
@fruch woops, forgot an `else` -- silly macro trickery got me again :(

**Comment by asottile** [comment]
Please don't merge this yet, it depends on changes made in https://github.com/python/cpython/pull/4797

**Comment by asottile** [comment]
I've rebased this now that #4797 is merged, should be ready for review again

File: Lib/test/test_shutil.py
Comment (vstinner) [inline Lib/test/test_shutil.py]: A whole temporary directory for a single file seems overkill. Why noy using NamedTemporaryFile instead? Or tempfile.mktemp() and then unlink it?

File: Lib/test/test_shutil.py
Comment (vstinner) [inline Lib/test/test_shutil.py]: I would prefer to see a check on the error code, or pass-through other exceptions.

File: Lib/test/test_shutil.py
Comment (vstinner) [inline Lib/test/test_shutil.py]: _has_functional_lchmod should be UPPER CASE. No need to make the constant private. Call it HAVE_FUNCTIONAL_LCHMOD?

File: Modules/posixmodule.c
Comment (vstinner) [inline Modules/posixmodule.c:2859]: Oh, there are two different error codes? Which one is raised on muslc?

Comment (asottile) [reply Modules/posixmodule.c:2859]: muslc raises `EOPNOTSUPP` (errno 95) though on my platform they were `#define`d to the same thing.  I simply picked the same errnos as `fchmodat` catches below -- I suspect implementations use `fchmodat` to implement `lchown` (or the other way around).

File: Lib/test/test_shutil.py
Comment (asottile) [reply Lib/test/test_shutil.py]: It's a file and a symlink -- then I'd need manual cleanup code (a `try-finally` + `unlink`) -- unless you feel strongly about this I'd rather leave it.  It's in test code after all and ephemeral.

File: Lib/test/test_shutil.py
Comment (asottile) [reply Lib/test/test_shutil.py]: done

File: Lib/test/test_shutil.py
Comment (asottile) [reply Lib/test/test_shutil.py]: done

File: Lib/test/test_shutil.py
Comment (asottile) [reply Lib/test/test_shutil.py]: Oh actually you're right I can probably get away with just a temp file. Was misremembering my implementation 😆

Comment (asottile) [reply Lib/test/test_shutil.py]: done!

**Comment by asottile** [comment]
@vstinner would love another review :) no rush though!

**Comment by asottile** [comment]
@vstinner friendly ping :)

**Comment by asottile** [comment]
Looking through [devguide](https://devguide.python.org/experts/) -- maybe @giampaolo can review the `shutil` changes?

File: Lib/test/test_shutil.py
Comment (TheKevJames) [inline Lib/test/test_shutil.py:49]: Should this be `mkstemp`? `mktemp` [seems to be deprecated](https://docs.python.org/2/library/tempfile.html#tempfile.mktemp).

Comment (asottile) [reply Lib/test/test_shutil.py:49]: This was copied from adjacent test (in the original PR).  It's not exactly that `mktemp` is deprecated, it's that it's _insecure_.  However in this case, the same security problem would exist since it's essentially:

1. `mkstemp`
2. unlink the file made by `mkstemp`
3. symlink to the location made by `mkstemp`

There's a race between 2 and 3 where something could create a file

Fortunately it _doesn't really matter_ since it's just a test.

**Comment by asottile** [comment]
Almost this branch's birthday -- rebased to resolve conflicts

**Comment by fruch** [comment]
@asottile anything I could do to help ?

**Comment by asottile** [comment]
@fruch iirc an approving review (even from an outsider) will move the state from [awaiting review] to [awaiting core review] which might (???) gather more attention from core devs

**Review by TheKevJames — APPROVED** [review]
Changeset LGTM, but I have not worked with this chunk of the cpython codebase before. Would be great to get a core contributor review as well!

**Comment by jdubost** [comment]
Dunno if this can help but i encountered the issue (https://bugs.python.org/issue31940) with a python-3.6 tox environment within an alpine image `python:2.7-alpine3.6` (+ python-3.6.5 installed manually)
Then i upgraded to a `python:3.6.7-alpine3.6` image (python-3.6.7 native) -> it solved the issue.

Since tox on a python2 environment works well on `python:3.6.7-alpine3.6`, i now use this image for all my tests.

**Comment by asottile** [comment]
@jdubost looks like alpine is working around this bug by manually turning off HAVE_LCHMOD:

```
python:3.6.7-alpine3.6$ grep HAVE_LCHMOD /usr/local/include/python3.6m/pyconfig.h 
/* #undef HAVE_LCHMOD */
```

```
python:3.6.1-alpine3.6$ grep HAVE_LCHMOD /usr/local/include/python3.6m/pyconfig.h 
#define HAVE_LCHMOD 1
```

It looks like @benjaminp forced this to off on all linuxes in 40caa05fa4d1810a1a6bfc34e0ec930c351089b7

which looks eerily similar to my _rejected_ patch in #4267 😭

**Comment by benjaminp** [comment]
Hmm, I didn't see any of this.

I don't understand the objection to a compile-time check. Whether you're targeting Linux or not is something you must know at compile time. It would be one thing if an operating system existed that sometimes returned `ENOSYS` from `lchmod`. But, I know of no such system.

**Comment by asottile** [comment]
unclear as well.  I guess since this is solved I can close this and we can resolve the linked issue.

I think some of the changes in the patch are useful so I'll make a separate bpo and propose them (notably, `os.stat` / `os.chmod` are now always present so the `lookup(...)` calls are unnecessary)

**Pull request closed without merging**
